truelearn/models/_knowledge.py

```diff
@@ -123,11 +124,114 @@ def clone(
+            "title": self.__title,
+            "description": self.__description,
+            "url": self.__url,
+        }
+
+
+class HistoryAwareKnowledgeComponent(KnowledgeComponent):
+    """A knowledge component that keeps a history about how it was updated."""
+
+    def __init__(
+        self,
+        *,
+        mean: float,
+        variance: float,
+        timestamp: Optional[float] = None,
+        title: Optional[str] = None,
+        description: Optional[str] = None,
+        url: Optional[str] = None,
+        history_limit: Optional[int] = None,
+        history: Optional[Deque[Tuple[float, float, Optional[float]]]] = None,
+    ) -> None:
+        """Init the KnowledgeComponent object.
+
+        Args:
+            mean:
+                A float indicating the mean of the knowledge component.
+            variance:
+                A float indicating the variance of the knowledge component.
+            timestamp:
+                A float indicating the POSIX timestamp of the last update of
+                the knowledge component.
+            title:
+                An optional string storing the title of the knowledge component.
+            description:
+                An optional string that describes the knowledge component.
+            url:
+                An optional string storing the url of the knowledge component.
+            history_limit:
+                A positive int that specifies the number of entries stored in the
+                history. If the limit is None, it means there is no limit.
+                Defaults to None.
+            history:
+                A queue that stores the update history of the knowledge component.
+                Each entry in the queue is a tuple (mean, variance, timestamp)
+                which records the mean and variance of this knowledge component
+                at the given timestamp.
+
+        Returns:
+            None.
+        """
+        super().__init__(
+            mean=mean,
+            variance=variance,
+            timestamp=timestamp,
+            title=title,
+            description=description,
+            url=url,
         )
 
+        if history is None:
+            history = collections.deque(maxlen=history_limit)
+        else:
+            if history.maxlen != history_limit:
+                history = collections.deque(history, maxlen=history_limit)
+        self.__history = history
+
+    @property
+    def history(self) -> Deque[Tuple[float, float, Optional[float]]]:
+        """The update history of the current knowledge component."""
+        return self.__history
+
+    def update(
+        self,
+        *,
+        mean: Optional[float] = None,
+        variance: Optional[float] = None,
+        timestamp: Optional[float] = None,
+    ) -> None:
+        self.__history.append((self.mean, self.variance, self.timestamp))
+
+        super().update(mean=mean, variance=variance, timestamp=timestamp)
+
+    def clone(
```

Comment:
docstring?

Reply:
What's the problem of this one?

Reply:
> What's the problem of this one?

There is no docstring for clone

Reply:
It inherits the docstring from `KnowledgeComponent`.